#!/usr/bin/python
import subprocess

dextract='/home/atimms/DEXTRACTOR/dextract'
bwa = '/tools/BioBuilds-2015.04/bin/bwa'
samtools = '/tools/BioBuilds-2015.04/bin/samtools-1.0'

bax_h5_files = ['m140619_203602_00127_c100649692550000001823117410071454_s1_p0.1.bax.h5', 'm140619_203602_00127_c100649692550000001823117410071454_s1_p0.2.bax.h5', 'm140619_203602_00127_c100649692550000001823117410071454_s1_p0.3.bax.h5', 'm140619_235515_00127_c100649692550000001823117410071455_s1_p0.1.bax.h5', 'm140619_235515_00127_c100649692550000001823117410071455_s1_p0.2.bax.h5', 'm140619_235515_00127_c100649692550000001823117410071455_s1_p0.3.bax.h5', 'm140620_212952_00127_c100650992550000001823117710071430_s1_p0.1.bax.h5', 'm140620_212952_00127_c100650992550000001823117710071430_s1_p0.2.bax.h5', 'm140620_212952_00127_c100650992550000001823117710071430_s1_p0.3.bax.h5', 'm140621_004905_00127_c100650992550000001823117710071431_s1_p0.1.bax.h5', 'm140621_004905_00127_c100650992550000001823117710071431_s1_p0.2.bax.h5', 'm140621_004905_00127_c100650992550000001823117710071431_s1_p0.3.bax.h5', 'm140621_040818_00127_c100650992550000001823117710071432_s1_p0.1.bax.h5', 'm140621_040818_00127_c100650992550000001823117710071432_s1_p0.2.bax.h5', 'm140621_040818_00127_c100650992550000001823117710071432_s1_p0.3.bax.h5', 'm140621_072731_00127_c100650992550000001823117710071433_s1_p0.1.bax.h5', 'm140621_072731_00127_c100650992550000001823117710071433_s1_p0.2.bax.h5', 'm140621_072731_00127_c100650992550000001823117710071433_s1_p0.3.bax.h5', 'm140621_104644_00127_c100650992550000001823117710071434_s1_p0.1.bax.h5', 'm140621_104644_00127_c100650992550000001823117710071434_s1_p0.2.bax.h5', 'm140621_104644_00127_c100650992550000001823117710071434_s1_p0.3.bax.h5', 'm140621_140847_00127_c100650992550000001823117710071435_s1_p0.1.bax.h5', 'm140621_140847_00127_c100650992550000001823117710071435_s1_p0.2.bax.h5', 'm140621_140847_00127_c100650992550000001823117710071435_s1_p0.3.bax.h5', 'm140621_172510_00127_c100650992550000001823117710071436_s1_p0.1.bax.h5', 'm140621_172510_00127_c100650992550000001823117710071436_s1_p0.2.bax.h5', 'm140621_172510_00127_c100650992550000001823117710071436_s1_p0.3.bax.h5', 'm140621_224814_00127_c100650992550000001823117710071437_s1_p0.1.bax.h5', 'm140621_224814_00127_c100650992550000001823117710071437_s1_p0.2.bax.h5', 'm140621_224814_00127_c100650992550000001823117710071437_s1_p0.3.bax.h5', 'm140622_020715_00127_c100650742550000001823117710071440_s1_p0.1.bax.h5', 'm140622_020715_00127_c100650742550000001823117710071440_s1_p0.2.bax.h5', 'm140622_020715_00127_c100650742550000001823117710071440_s1_p0.3.bax.h5', 'm140622_052626_00127_c100650742550000001823117710071441_s1_p0.1.bax.h5', 'm140622_052626_00127_c100650742550000001823117710071441_s1_p0.2.bax.h5', 'm140622_052626_00127_c100650742550000001823117710071441_s1_p0.3.bax.h5', 'm140622_084548_00127_c100650742550000001823117710071442_s1_p0.1.bax.h5', 'm140622_084548_00127_c100650742550000001823117710071442_s1_p0.2.bax.h5', 'm140622_084548_00127_c100650742550000001823117710071442_s1_p0.3.bax.h5', 'm140622_120446_00127_c100650742550000001823117710071443_s1_p0.1.bax.h5', 'm140622_120446_00127_c100650742550000001823117710071443_s1_p0.2.bax.h5', 'm140622_120446_00127_c100650742550000001823117710071443_s1_p0.3.bax.h5', 'm140626_202900_00127_c100650862550000001823117710071497_s1_p0.1.bax.h5', 'm140626_202900_00127_c100650862550000001823117710071497_s1_p0.2.bax.h5', 'm140626_202900_00127_c100650862550000001823117710071497_s1_p0.3.bax.h5', 'm140628_043719_00127_c100650742550000001823117710071447_s1_p0.1.bax.h5', 'm140628_043719_00127_c100650742550000001823117710071447_s1_p0.2.bax.h5', 'm140628_043719_00127_c100650742550000001823117710071447_s1_p0.3.bax.h5', 'm140628_075632_00127_c100650702550000001823117710071480_s1_p0.1.bax.h5', 'm140628_075632_00127_c100650702550000001823117710071480_s1_p0.2.bax.h5', 'm140628_075632_00127_c100650702550000001823117710071480_s1_p0.3.bax.h5', 'm140628_111818_00127_c100650702550000001823117710071481_s1_p0.1.bax.h5', 'm140628_111818_00127_c100650702550000001823117710071481_s1_p0.2.bax.h5', 'm140628_111818_00127_c100650702550000001823117710071481_s1_p0.3.bax.h5', 'm140628_143458_00127_c100650702550000001823117710071482_s1_p0.1.bax.h5', 'm140628_143458_00127_c100650702550000001823117710071482_s1_p0.2.bax.h5', 'm140628_143458_00127_c100650702550000001823117710071482_s1_p0.3.bax.h5', 'm140628_195644_00127_c100650702550000001823117710071483_s1_p0.1.bax.h5', 'm140628_195644_00127_c100650702550000001823117710071483_s1_p0.2.bax.h5', 'm140628_195644_00127_c100650702550000001823117710071483_s1_p0.3.bax.h5', 'm140628_231540_00127_c100650702550000001823117710071484_s1_p0.1.bax.h5', 'm140628_231540_00127_c100650702550000001823117710071484_s1_p0.2.bax.h5', 'm140628_231540_00127_c100650702550000001823117710071484_s1_p0.3.bax.h5', 'm140629_023453_00127_c100650702550000001823117710071485_s1_p0.1.bax.h5', 'm140629_023453_00127_c100650702550000001823117710071485_s1_p0.2.bax.h5', 'm140629_023453_00127_c100650702550000001823117710071485_s1_p0.3.bax.h5', 'm140629_055406_00127_c100650702550000001823117710071486_s1_p0.1.bax.h5', 'm140629_055406_00127_c100650702550000001823117710071486_s1_p0.2.bax.h5', 'm140629_055406_00127_c100650702550000001823117710071486_s1_p0.3.bax.h5']

combined_fq_file = 'zf_sanger.pacbio.fq.gz'

zv10_scaffold_fa = 'zv10_scaffolds.fa'
pacbio_scaffold_bam = 'zf_sanger.zv10_scaffolds.bwa.bam'

# bax_h5_files = ['m140619_203602_00127_c100649692550000001823117410071454_s1_p0.1.bax.h5']
##covert baxh5 to fasta format
# for bax_h5 in bax_h5_files:
# 	fasta_file = bax_h5.rsplit('.',2)[0] + '.fasta'
# 	with open(fasta_file, "w") as fa_fh:
# 		grep_cmd = subprocess.Popen([dextract, bax_h5], stdout=fa_fh)
# 		grep_cmd.wait()

## manually combine fasta all files into file: 'zf_sanger.pacbio.fa' i.e. cat *fasta > zf_sanger.pacbio.fa

##covert baxh5 to fastq format
for bax_h5 in bax_h5_files:
	fastq_file = bax_h5.rsplit('.',2)[0] + '.fastq'
	with open(fastq_file, "w") as fa_fh:
		grep_cmd = subprocess.Popen([dextract, '-q', bax_h5], stdout=fa_fh)
		grep_cmd.wait()
	gzip_cmd = subprocess.Popen(['gzip', fastq_file])
	gzip_cmd.wait()

## manually combine fasta all files into file i.e. cat *fastq.gz > zf_sanger.pacbio.fq.gz


##map using bwa against scaffolds
# bwa_pe = subprocess.Popen([bwa, 'mem', '-t', '20', '-x', 'pacbio',  zv10_scaffold_fa, combined_fq_file], stdout=subprocess.PIPE)
# st_sam_bam_pe = subprocess.Popen([samtools, 'view', '-b', '-@', '5', '-o', pacbio_scaffold_bam, '-'], stdin=bwa_pe.stdout)
# st_sam_bam_pe.wait()